sashimi-webapp/src/logic/filemanager/data/file.js

```diff
@@ -136,23 +146,75 @@ File.prototype.move = function move(destFolder) {
 };
 
 /**
+ * Make a copy of a file in the current Folder
+ *
+ * @return {Promise}
+ */
+File.prototype.copy = function copy() {
+  return storage.copyFile(this.id)
+  .then((dbCopiedFile) => {
+    const copiedFile = new File(dbCopiedFile.file_id, dbCopiedFile.file_name, dbCopiedFile.file_path, this);
+    this.parentFolder.childFileList.push(copiedFile);
+    idMap.addFileToMap(copiedFile.id, copiedFile);
+  });
+};
+
+/**
  * Rename file
  *
  * @param {String} newFileName
  * @return {Promise}
  */
 File.prototype.rename = function rename(newFileName) {
-  return new Promise((resolve, reject) => {
-    if (hasSameFileName.call(this, newFileName)) {
-      reject(ERROR_SAME_FILE_NAME);
-    }
+  if (!(this instanceof File)) {
+    throw new Error(ERROR_NOT_FILE_INSTANCE);
+  }
 
-    resolve();
-  })
-  .then(() => storage.renameFile(this.id, newFileName))
-  .then(() => {
-    const oldFileName = this.name;
-    this.name = newFileName;
-    this.path = this.path.replace(oldFileName, newFileName);
+  newFileName = newFileName.trim();
+  if (!newFileName) {
+    return Promise.reject(new Error(ERROR_EMPTY_STRING));
+  } else if (this.name === newFileName) {
+    return Promise.resolve();
+  } else if (newFileName.match(ILLEGAL_CHARACTERS)) {
+    return Promise.reject(new Error(ERROR_CONTAIN_ILLEGAL_CHARACTERS));
+  } else if (hasSameFileName.call(this, newFileName)) {
+    return Promise.reject(new Error(ERROR_SAME_FILE_NAME));
+  } else {
+    return storage.renameFile(this.id, newFileName)
+    .then(() => {
+      const oldFileName = this.name;
+      this.name = newFileName;
+      this.path = this.path.replace(oldFileName, newFileName);
+    });
+  }
+};
+
+/**
+ * Download file
+ * (Code transferred from Filemanger.vue)
+ *
+ * @return {Promise}
+ */
+File.prototype.download = function download() {
+  return this.load()
+  .then((data) => {
+    const fileName = this.name.concat('.md');
```

Comment:
The `.md` file extension is used here, when the user download a file into their system.

- This is done so that the user can directly open the file with a markdown editor.
- It is also used to prevent user from downloading file that contain bad extension such as `.exe`